torch/backends/_nnapi/serializer.py

```diff
@@ -0,0 +1,1131 @@
+
+        return serialized_values, serialized_value_data
+
+
+    @staticmethod
+    def serialize_ints(ints):
+        return struct.pack("i" * len(ints), *ints)
+
+
+    ADDER_MAP = {
+            "prim::GetAttr": lambda self, node:
+                self.add_getattr(node),
+            "prim::Constant": lambda self, node:
+                self.add_constant_node(node),
+            "prim::ListConstruct": lambda self, node:
+                self.add_list_construct(node),
+            "aten::quantize_per_tensor": lambda self, node:
+                self.add_quantize(node),
+            "aten::dequantize": lambda self, node:
+                self.add_dequantize(node),
+            "aten::add": lambda self, node:
+                self.add_add_sub_op(node, NNAPI_OperationCode.ADD, NNAPI_FuseCode.FUSED_NONE),
+            "aten::sub": lambda self, node:
+                self.add_add_sub_op(node, NNAPI_OperationCode.SUB, NNAPI_FuseCode.FUSED_NONE),
+            "aten::mul": lambda self, node:
+                self.add_pointwise_simple_binary_broadcast_op(node, NNAPI_OperationCode.MUL),
+            "aten::relu": lambda self, node:
+                self.add_pointwise_simple_unary_op(node, NNAPI_OperationCode.RELU),
+            "aten::sigmoid": lambda self, node:
+                self.add_pointwise_simple_unary_op(node, NNAPI_OperationCode.LOGISTIC),
+            "aten::max_pool2d": lambda self, node:
+                self.add_pool2d_node(node, NNAPI_OperationCode.MAX_POOL_2D),
+            "aten::upsample_nearest2d": lambda self, node:
+                self.add_upsample_nearest2d(node),
+            "aten::prelu": lambda self, node:
+                self.add_prelu_op(node),
+            "aten::_convolution": lambda self, node:
+                self.add_conv(node),
+            "quantized::conv2d": lambda self, node:
+                self.add_qconv2d(node, NNAPI_FuseCode.FUSED_NONE),
+            "quantized::conv2d_relu": lambda self, node:
+                self.add_qconv2d(node, NNAPI_FuseCode.FUSED_RELU),
+            "quantized::add": lambda self, node:
+                self.add_qadd(node, NNAPI_OperationCode.ADD, NNAPI_FuseCode.FUSED_NONE),
+            "quantized::add_relu": lambda self, node:
+                self.add_qadd(node, NNAPI_OperationCode.ADD, NNAPI_FuseCode.FUSED_RELU),
+    }
+
+
+    def add_node(self, node):
+        adder = self.ADDER_MAP.get(node.kind())
+        if not adder:
+            print(node)
+            raise Exception("Unsupported node type: %r" % node.kind())
+        adder(self, node)
+
+
+    def add_getattr(self, node):
+        assert node.inputsSize() == 1
+        assert node.outputsSize() == 1
+        obj_ctype, obj = self.get_constant_value(node.inputsAt(0))
+        assert str(obj_ctype).startswith("__torch__.")
+        name = node.s("name")
+        value = getattr(obj, name)
+        output = node.outputsAt(0)
+        ctype = output.type()
+        self.add_constant_value(output, ctype, value)
```

Comment:
Good idea to freeze the model! So in nnapi models, there's no concept of "parameters"? The weights and bias are all "constants"? Does that mean it does not have training capability?

Reply:
NNAPI is only for inference.

Reply:
> > Would this model only be used in platforms with nnapi available (Android)?
> 
> Yes. (Though it is possible in theory to build the NNAPI CPU implementation for Linux and run the model directly, which is how I've been testing.)

Does that mean we may keep multiple versions of a model?
For Android, we deploy an nnapi model.
For iOS, I'm wondering if there is corresponding API layer (CoreML?)

Reply:
@iseeyuan for Metal, I believe it's integrated as a separate backend instead. I'm also curious to know why NNAPI isn't implemented as a separate backend. One thought I had is that PyTorch represents image Tensors as NCHW whereas NNAP (based on @dreiss diff summary) requires them to be in NHWC format, so implementing it as a separate backend may have required the conversion on every operator entry/exit? Not sure.